text/trace/0220-messaging-semantic-conventions-span-structure.md

```diff
@@ -0,0 +1,484 @@
+code.
+
+##### Pull-based scenarios
+
+"Receive" spans SHOULD be created for operations of passing messages to the
+application when those operations are initiated by the application code.
+
+##### General considerations
+
+"Deliver" or "Receive" spans MUST NOT be created for messages which are not
+forwarded to the application, but are pre-fetched or cached by messaging
+libraries or SDKs.
+
+A single "Deliver" or "Receive" span can account for a single message, for
+multiple messages (in case messages are passed for processing as batches), or
+for no message at all (if it is signalled that no messages were received).  For
+each message it accounts for, the "Deliver" or "Receive" span SHOULD link to
+the "Create" span for the message.
+
+#### Settlement spans
+
+"Settle" spans or events SHOULD be created for every manually or automatically
+triggered settlement operation. A single "Settle" span can account for a
+single message or for multiple messages (in case messages are passed for
+settling as batches). For each message it accounts for, the "Settle" span
+MAY link to the "Create" span for the message.
+
+## Examples
+
+### Single message producer, single message push-based consumer
+
+```mermaid
+flowchart LR;
+  subgraph PRODUCER
+  PM1[Publish m1]
+  end
+  subgraph CONSUMER
+  DM1[Deliver m1]
+  end
+  PM1-. link .->DM1;
+```
+
+### Single message producer, single message push-based consumer with manual settlement
+
+```mermaid
+flowchart LR;
+  subgraph PRODUCER
+    direction TB
+    PM1[Publish m1]
+  end
+  subgraph CONSUMER
+    direction TB
+    DM1[Deliver m1]-->S1[Settle m1]
+  end
+  PM1-. link .->DM1;
+```
+
+### Single message producer, single message push-based consumer with auto-settlement
+
+```mermaid
+flowchart LR;
+  subgraph PRODUCER
+  direction TB
+  PM1[Publish m1]
+  end
+  subgraph CONSUMER
+  direction TB
+  A[Ambient]-->DM1[Deliver m1]
+  A-->S1[Settle m1]
+  end
+  PM1-. link .->DM1;
+```
+
+### Batch message producer with "Create" spans, single message pull-based consumer
+
+```mermaid
+flowchart LR;
+  subgraph PRODUCER
+  direction TB
+  P[Publish]-->CM1[Create m1]
+  P-->CM2[Create m2]
+  end
+  subgraph CONSUMER
+  direction TB
+  RM1[Receive m1]
+  RM2[Receive m1]
+  end
+  CM1-. link .->RM1;
+  CM2-. link .->RM2;
+```
+
+### Batch message producer, single message push-based consumer
+
+```mermaid
+flowchart LR;
+  subgraph PRODUCER
+  direction TB
+  P[Publish]
+  end
+  subgraph CONSUMER
+  direction TB
+  DM1[Deliver m1]
+  DM2[Deliver m1]
+  end
+  P-. link .->DM1;
+  P-. link .->DM2;
+```
+
+### Batch message producer with "Create" spans populated before publish, single message pull-based consumer
+
+```mermaid
+flowchart LR;
+  subgraph PRODUCER
+  direction TB
+  A[Ambient]-->CM1[Create m1]
+  A-->CM2[Create m2]
+  A-->P[Publish]
+  end
+  subgraph CONSUMER
+  direction TB
+  RM1[Receive m1]
+  RM2[Receive m1]
+  end
+  CM1-. link .->RM1;
+  CM2-. link .->RM2;
+```
```

Comment:
shoud this show that "Create" spans are CHILD_OF the "Publish" one?

Reply:
No, as the header for this example says: _Batch message producer with "Create" spans populated **before** publish_.

This is to support scenarios where users want full control and provide a creation context on their own, which is then attached to the message.

Reply:
I still don't get it. From the diagram it anyway seems that the consumer has received the messages so there is a link between "Create" and "Receive" spans but "Publish" is not correlated with the "Create" yet which looks really strange. If you are referring to something like " ... spans populated before publish." then I would expect that if not published, we should not have "Receive" spans on the right because not published, means not received yet.